spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -2238,11 +2240,14 @@ protected void doInTransactionWithoutResult(TransactionStatus status) {
 
 		private List<ConsumerRecord<K, V>> createRecordList(final ConsumerRecords<K, V> records) {
 			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
-			List<ConsumerRecord<K, V>> list = new LinkedList<>();
+			@SuppressWarnings("unchecked") ConsumerRecord<K, V>[] recordsArray =
+					(ConsumerRecord<K, V>[]) Array.newInstance(ConsumerRecord.class, records.count());
+			int index = 0;
 			while (iterator.hasNext()) {
-				list.add(iterator.next());
+				recordsArray[index] = iterator.next();
+				index += 1;
 			}
-			return list;
+			return List.of(recordsArray);
```

Comment:
No, that's not what I was talking about.
To minimize braking change, please, let's consider to use here `new ArrayList()` instead.
And revert that change in the `RecordFilterStrategy`.

The point is that mostly this list is going to be used as is without any extra intermediate objects to mitigate its immutability. And if we look at this from the high point of view, it is really up to target application to deal with data properly, while we should let them to do with this list whatever they want and need.